### DISCUSSION TITLE
Ethernet & USB

### DISCUSSION THREAD

**Discussion by mlewus** [terjeio/grblHAL#310] [Q&A]
I am setting up a machine with grblHAL running on Teensy 4.1 (i.MXRT1062). I am planning on compiling with ethernet enabled. With the code built that way can I also use the USB port for sending gcode? Not at the same time, but with the same build. Thanks!

**Comment by phil-barrett** [comment]
Yes,  That will work just fine.  In fact, you will need USB in order to do the initial settings set up and get the ip address.

**Reply by stevempotter** [reply]
What about at the same time? I am also going to be using the Ethernet port to communicate with a computer, and the Teensy's USB to get commands transmitted to an ESP32 by my wireless pendant (and send acknowledgements and other info). I would like the pendant to be able to interrupt ongoing motions, e.g., by sending a Halt or altering the feedrate.

**Reply by stevempotter** [reply]
Or is it better to use the I2C port on your breakout board instead of USB? Is one faster than the other?

**Reply by terjeio** [reply]
USB is fastest, I2C can be faster than UART depending on the transmission speed. Currently only device mode is available for USB and a network connection disables the USB connection. Host mode is possible and the Pro version of Phil's breakout board has a connector for that. Code needs to be added to support it though.

The [keypad plugin](https://github.com/grblHAL/Plugin_I2C_keypad) works with I2C transmission and is currently the only input channel that can be used for input in paralell with the main stream regardless if it is via UART, USB or a network stream.

I have recently added support for the [HC-05 Bluetooth module](https://github.com/grblHAL/Plugins_Bluetooth) (via the UART port), I plan to update the underlying stream code so it can be used as a comms channel for MPGs - possibly by defining a MPG protocol on top of Modbus and/or as an alternative to I2C comms for the keypad plugin.

Please raise a new issue or start a discussion at the [new repo](https://github.com/grblHAL) to continue this discussion as the OPs question has been answered.

**Reply by phil-barrett** [reply]
Thanks, Terje.
